File: clients/android/StellarChat/app/src/main/java/com/stellarmls/chat/nostr/InvitationTransport.kt
package com.stellarmls.chat.nostr

import com.stellarmls.chat.crypto.GroupCrypto
import com.stellarmls.chat.crypto.KeyManager
import com.stellarmls.chat.crypto.NostrEvent
import com.stellarmls.chat.crypto.NostrEventBuilder
import com.stellarmls.chat.model.BootstrapPayload
import com.stellarmls.chat.model.PendingInvitation
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.json.JSONArray
import org.json.JSONObject
import java.util.Date
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Kind 34113 invitation send/receive over Nostr relays.
 * Still listens for legacy kind 24113 events during migration.
 * Uses X25519 ECDH + AES-256-GCM for invitation encryption.
 */
class InvitationTransport(private val keyManager: KeyManager) {
    companion object {
        private const val PRIMARY_KIND = 34113
        private const val LEGACY_KIND = 24113
        private const val PRIMARY_INBOX_FILTER_KEY = "#d"
        private const val PRIMARY_INBOX_EVENT_TAG_KEY = "d"
        private const val SECONDARY_INBOX_EVENT_TAG_KEY = "t"
        private const val LEGACY_INBOX_EVENT_TAG_KEY = "sep_inbox"
        private const val PRIMARY_INBOX_TAG_PREFIX = "sep-inbox:"

        fun subscriptionFilters(inboxTag: String): List<JSONObject> = listOf(
            JSONObject().apply {
                put("kinds", JSONArray().put(PRIMARY_KIND))
                put(PRIMARY_INBOX_FILTER_KEY, JSONArray().put(PRIMARY_INBOX_TAG_PREFIX + inboxTag))
            },
            JSONObject().apply {
                put("kinds", JSONArray().put(PRIMARY_KIND))
                put("#t", JSONArray().put(inboxTag))
            },
            JSONObject().apply {
                put("kinds", JSONArray().put(LEGACY_KIND))
                put("#t", JSONArray().put(inboxTag))
            }
        )

        fun eventTags(recipientInboxTag: String): List<List<String>> = listOf(
            // Use a parameterized-replaceable `d` tag on an addressable kind so
            // relays can retain and query invitations across reconnects.
            listOf(PRIMARY_INBOX_EVENT_TAG_KEY, PRIMARY_INBOX_TAG_PREFIX + recipientInboxTag),
            listOf(SECONDARY_INBOX_EVENT_TAG_KEY, recipientInboxTag),
            listOf(LEGACY_INBOX_EVENT_TAG_KEY, recipientInboxTag),
            listOf("sep_version", "1")
        )
    }

    private val connections = mutableListOf<NostrRelayConnection>()
    private val subscriptionJobs = ConcurrentHashMap<String, Job>()
    private val scope = CoroutineScope(Dispatchers.IO)

    var onInvitation: ((PendingInvitation) -> Unit)? = null
    var onError: ((String) -> Unit)? = null

    fun connect(relayURLs: List<String>) {
        for (url in relayURLs) {
            val conn = NostrRelayConnection(url)
            conn.connect()
            connections.add(conn)
        }
        if (com.stellarmls.chat.BuildConfig.DEBUG) {
            android.util.Log.d("InvitationTransport", "connect relays=$relayURLs")
        }
    }

    fun disconnect() {
        subscriptionJobs.values.forEach { it.cancel() }
        subscriptionJobs.clear()
        connections.forEach { it.disconnect() }
        connections.clear()
    }

    /** Subscribe to inbox for incoming invitations. */
    fun subscribeToInbox(inboxTag: String, privateKey: X25519PrivateKeyParameters) {
        val subID = "inbox-${UUID.randomUUID().toString().take(8)}"

        val filters = subscriptionFilters(inboxTag)
        if (com.stellarmls.chat.BuildConfig.DEBUG) {
            android.util.Log.d("InvitationTransport", "subscribeToInbox inboxTag=$inboxTag filters=$filters")
        }

        for (conn in connections) {
            filters.forEachIndexed { index, filter ->
                val filterSubID = "$subID-$index"
                val job = conn.subscribe(filterSubID, filter)
                    .onEach { event -> handleInvitationEvent(event, privateKey) }
                    .launchIn(scope)
                subscriptionJobs["$filterSubID-${conn.hashCode()}"] = job
            }
        }
    }

    /** Send an invitation to a recipient via their inbox. */
    fun sendInvitation(
        payload: BootstrapPayload,
        recipientKeyAgreementPubkey: ByteArray,
        keyManager: KeyManager
    ) {
        val payloadJson = payload.toJson()
        val envelopeJson = GroupCrypto.encryptInvitation(
            payload = payloadJson.toByteArray(),
            recipientKeyAgreementPubkey = recipientKeyAgreementPubkey,
            senderKeyManager = keyManager
        )

        // Base64 encode the envelope
        val contentBase64 = android.util.Base64.encodeToString(
            envelopeJson.toByteArray(),
            android.util.Base64.NO_WRAP
        )

        val recipientInboxTag = GroupCrypto.hiddenInboxTag(recipientKeyAgreementPubkey)

        val tags = eventTags(recipientInboxTag)

        val event = NostrEventBuilder.build(
            kind = PRIMARY_KIND,
            tags = tags,
            content = contentBase64,
            keyManager = keyManager
        )

        for (conn in connections) {
            conn.publish(event)
        }
    }

    private fun handleInvitationEvent(event: NostrEvent, privateKey: X25519PrivateKeyParameters) {
        try {
            if (com.stellarmls.chat.BuildConfig.DEBUG) {
                android.util.Log.d(
                    "InvitationTransport",
                    "Received event id=${event.id.take(12)} kind=${event.kind} tags=${event.tags}"
                )
            }
            // Base64 decode the content to get the sealed envelope JSON
            val envelopeJson = String(
                android.util.Base64.decode(event.content, android.util.Base64.NO_WRAP)
            )

            // Decrypt using X25519 ECDH
            val payloadBytes = GroupCrypto.decryptInvitation(envelopeJson, privateKey)
            val payload = BootstrapPayload.fromJson(String(payloadBytes))

            val invitation = PendingInvitation(
                id = event.id,
                payload = payload,
                receivedAt = Date(event.displayMilliseconds)
            )

            onInvitation?.invoke(invitation)
        } catch (e: Exception) {
            if (com.stellarmls.chat.BuildConfig.DEBUG) {
                android.util.Log.d(
                    "InvitationTransport",
                    "Ignored event ${event.id.take(12)}: ${e.message}"
                )
            }
        }
    }
}

File: clients/ios/StellarChat/StellarChat/Nostr/InvitationTransport.swift
import CryptoKit
import Foundation
import SwiftMLS

/// InvitationTransport: handles kind 34113 (inbox-addressed, per-recipient).
/// Do NOT use this transport for kind 44114 (group broadcast) — use NostrMessageTransport.
/// Still listens for legacy kind 24113 events during migration.
@Observable
final class InvitationTransport {
    private static let primaryKind = 34113
    private static let legacyKind = 24113
    private static let primaryInboxFilterKey = "#d"
    private static let primaryInboxEventTagKey = "d"
    private static let secondaryInboxEventTagKey = "t"
    private static let legacyInboxEventTagKey = "sep_inbox"
    private static let primaryInboxTagPrefix = "sep-inbox:"

    private var connections: [URL: NostrRelayConnection] = [:]
    private var subscriptionTask: Task<Void, Never>?

    /// Called when an invitation is successfully decrypted.
    var onInvitation: ((PendingInvitation) -> Void)?
    /// Called when a rekey envelope is successfully decrypted.
    var onRekeyEnvelope: ((SEPRekeyEnvelope) -> Void)?
    /// Called on transport-level errors.
    var onError: ((String) -> Void)?

    static func subscriptionFilters(inboxTag: String) -> [[String: Any]] {
        [
            [
                "kinds": [primaryKind],
                primaryInboxFilterKey: [primaryInboxTagPrefix + inboxTag],
            ],
            [
                "kinds": [primaryKind],
                "#t": [inboxTag],
            ],
            [
                "kinds": [legacyKind],
                "#t": [inboxTag],
            ],
        ]
    }

    static func eventTags(recipientInboxTag: String) -> [[String]] {
        [
            // Use a parameterized-replaceable `d` tag on an addressable kind so
            // relays can retain and query invitations across reconnects.
            [primaryInboxEventTagKey, primaryInboxTagPrefix + recipientInboxTag],
            [secondaryInboxEventTagKey, recipientInboxTag],
            [legacyInboxEventTagKey, recipientInboxTag],
            ["sep_version", "1"],
        ]
    }

    func connect(to urls: [URL]) async {
        for url in urls {
            if connections[url] == nil {
                let conn = NostrRelayConnection(url: url)
                connections[url] = conn
                await conn.connect()
            }
        }
        #if DEBUG
        print("[Invite] connect relays=\(urls.map(\.absoluteString))")
        #endif
    }

    func disconnect() async {
        subscriptionTask?.cancel()
        subscriptionTask = nil
        for conn in connections.values {
            await conn.disconnect()
        }
        connections.removeAll()
    }

    /// Subscribe to invitation events addressed to our inbox.
    func subscribeToInbox(
        inboxTag: String,
        privateKey: Curve25519.KeyAgreement.PrivateKey
    ) {
        subscriptionTask?.cancel()

        let subID = "inbox-\(inboxTag)"

        subscriptionTask = Task { [weak self] in
            guard let self else { return }
            let filters = Self.subscriptionFilters(inboxTag: inboxTag)
            await withTaskGroup(of: Void.self) { group in
                for (_, conn) in self.connections {
                    for (index, filter) in filters.enumerated() {
                        group.addTask {
                            let filterSubID = "\(subID)-\(index)"
                            let stream = await conn.subscribe(subscriptionID: filterSubID, filter: filter)
                            for await event in stream {
                                guard !Task.isCancelled else { break }
                                self.handleInvitationEvent(event, privateKey: privateKey)
                            }
                        }
                    }
                }
            }
        }
    }

    /// Send an invitation to a specific recipient.
    func sendInvitation(
        payload: BootstrapPayload,
        recipientKeyAgreementPubkey: Data,
        keyManager: KeyManager
    ) async throws {
        let payloadData = try JSONEncoder().encode(payload)

        let envelope = try GroupCrypto.encryptInvitation(
            payloadData,
            recipientKeyAgreementPubkey: recipientKeyAgreementPubkey,
            senderSigningKey: keyManager.ed25519SigningKey
        )
        let envelopeData = try JSONEncoder().encode(envelope)
        let content = envelopeData.base64EncodedString()

        let recipientInboxTag = GroupCrypto.hiddenInboxTag(
            recipientPublicKey: recipientKeyAgreementPubkey
        )

        let tags = Self.eventTags(recipientInboxTag: recipientInboxTag)
        #if DEBUG
        print("[Invite] send recipientPubkey=\(recipientKeyAgreementPubkey.prefix(12).map { String(format: "%02x", $0) }.joined()) recipientInboxTag=\(recipientInboxTag) kind=\(Self.primaryKind) tags=\(tags)")
        #endif

        let event = try NostrEvent.build(
            kind: Self.primaryKind,
            tags: tags,
            content: content,
            keyManager: keyManager
        )

        let conns = Array(connections.values)
        let accepted = await withTaskGroup(of: Bool.self) { group in
            for conn in conns {
                group.addTask {
                    do {
                        let ok = try await conn.publishAndAwaitOK(event: event)
                        #if DEBUG
                        print("[Invite] publishAndAwaitOK: relay ok=\(ok) eventID=\(event.id.prefix(12))")
                        #endif
                        return ok
                    } catch {
                        #if DEBUG
                        print("[Invite] publishAndAwaitOK failed: \(error)")
                        #endif
                        return false
                    }
                }
            }
            var any = false
            for await result in group {
                if result { any = true }
            }
            return any
        }

        if !accepted {
            throw ChatError.relayPublishFailed
        }
    }

    /// Publish a pre-built event to all connected relays (concurrently).
    /// - Precondition: event.kind must be 34113 (inbox). Use NostrMessageTransport for 44114.
    func publishToRelays(_ event: NostrEvent, relayURLs: [URL] = []) async throws {
        assert(event.kind == Self.primaryKind || event.kind == Self.legacyKind,
               "InvitationTransport received non-inbox event kind \(event.kind) — use NostrMessageTransport for group messages")
        if connections.isEmpty && !relayURLs.isEmpty {
            #if DEBUG
            print("[Invite] publishToRelays: no connections — connecting to \(relayURLs.count) relays")
            #endif
            await connect(to: relayURLs)
        }
        let conns = Array(connections.values)
        #if DEBUG
        print("[Invite] publishToRelays: publishing to \(conns.count) connections")
        #endif
        guard !conns.isEmpty else {
            throw ChatError.relayPublishFailed
        }
        let accepted = await withTaskGroup(of: Bool.self) { group in
            for conn in conns {
                group.addTask {
                    do {
                        return try await conn.publishAndAwaitOK(event: event)
                    } catch {
                        #if DEBUG
                        print("[Invite] publishToRelays: relay publish failed: \(error)")
                        #endif
                        return false
                    }
                }
            }
            var any = false
            for await result in group {
                if result { any = true }
            }
            return any
        }
        #if DEBUG
        print("[Invite] publishToRelays: eventID=\(event.id.prefix(12)) kind=\(event.kind) relays=\(conns.count) accepted=\(accepted)")
        #endif
        if !accepted {
            throw ChatError.relayPublishFailed
        }
    }

    // MARK: - Private

    private func handleInvitationEvent(
        _ event: NostrEvent,
        privateKey: Curve25519.KeyAgreement.PrivateKey
    ) {
        #if DEBUG
        let tagSummary = event.tags.map { $0.joined(separator: ":") }.joined(separator: ",")
        print("[Invite] Received event id=\(event.id.prefix(12)) kind=\(event.kind) tags=\(tagSummary)")
        #endif
        guard let envelopeData = Data(base64Encoded: event.content) else {
            #if DEBUG
            print("[Invite] Base64 decode failed for event \(event.id.prefix(12))")
            #endif
            onError?("Failed to decode invitation base64")
            return
        }
        guard let envelope = try? JSONDecoder().decode(SealedEnvelope.self, from: envelopeData) else {
            #if DEBUG
            print("[Invite] Envelope decode failed for event \(event.id.prefix(12))")
            #endif
            onError?("Failed to decode invitation envelope")
            return
        }

        do {
            let payloadData = try GroupCrypto.decryptInvitation(envelope, privateKey: privateKey)
            // Try to parse as rekey envelope first, then as invitation
            if let rekeyEnvelope = try? JSONDecoder().decode(SEPRekeyEnvelope.self, from: payloadData),
               rekeyEnvelope.type == SEPRekeyEnvelope.messageType {
                onRekeyEnvelope?(rekeyEnvelope)
            } else {
                let payload = try JSONDecoder().decode(BootstrapPayload.self, from: payloadData)
                let invitation = PendingInvitation(
                    id: event.id,
                    payload: payload,
                    receivedAt: Date(timeIntervalSince1970: TimeInterval(event.displayMilliseconds) / 1000.0)
                )
                onInvitation?(invitation)
            }
        } catch {
            #if DEBUG
            print("[Invite] Ignored event \(event.id.prefix(12)): \(error.localizedDescription)")
            #endif
        }
    }
}
